x/sync/README.md

```diff
@@ -1,3 +1,59 @@
+
+If a client requests a range proof for range [`start`, `end`] but only receives a range proof for [`start`, `end'`] where `end'` < `end` it recognizes that it must still fetch all of the keys in [`end'`, `end`]. It repeatedly requests range proofs for chunks of the remaining key range until it has all of the key-value pairs in [`start`, `end`]. Note that the client may split the remaining key range into chunks and fetch chunks of key-value pairs in parallel.
+
+The database may be changing as the client is syncing. The sync client can be notified that the root ID of the database it's trying to sync to has changed. Detecting that the root ID to sync to has changed is done outside this package. If this occurs, the key-value pairs the client has learned about via range proofs may no longer be up to date.
+
+We use change proofs to correct the out of date key-value pairs. When the sync client is notified that the root ID to sync to has changed, it requests a change proof from a server for a given key range. For example, if a client has the key-value pairs in range [`start`, `end`] that were in the database when it had `rootID`, then it will request a change proof that provides all of the key-value changes in range [`start`, `end`] from the database version with root ID `rootID` to the database version with root ID `newRootID`. The client verifies the change proof, and if it's valid, it applies the changes to its database. If it's not, the client drops the proof and requests the proof from another server.
+
+A server needs to have history in order to serve a change proof. Namely, it needs to know all of the database changes between two roots. If the server does not have sufficient history to generate a change proof, it will send a range proof instead. The client will verify and apply the range proof. Note that change proofs, like range proofs, may not contain all of the key-value pairs in the requested range. This is OK because as mentioned above, the client tracks the root ID associated with each range of key-value pairs it has, so it knows which key-value pairs are out of date. Similar to range proofs, if a client requests the changes in range [`start`, `end`], but the server replies with all of the changes in [`start`, `end'`] for some `end'` < `end`, the client will repeatedly request change proofs until it gets remaining key-value pairs (namely in [`end'`, `end`]). 
```

Comment:
Does the client know that it sent a `RangeProof`? Seems like it has to know, because it should implicitly delete anything that isn't present in the `RangeProof` that it currently has.

I think you can avoid the repetition here and just mention that the `ChangeProof` might be incomplete, and the same algorithm is applied to fetch more changes as mentioned earlier. Actually it looks like you talk about that in the next paragraph.

Reply:
> Seems like it has to know, because it should implicitly delete anything that isn't present in the RangeProof that it currently has.

It doesn't necessarily need to delete everything in the old range proof, only the key-value pairs that are covered by the new range proof.

For example, if the client has a range proof for [`start_1`, `end_1`] at `root_hash_1`, and it gets a range proof for [`start_1`, `end_2`] at `root_hash_2` where `end_2` < `end_1` and `root_hash_2` is after `root_hash_1`, then it can still keep the key-values in [`end_2`, `end_1`] from the revision at `root_hash_1`. It'll just have to update those via change proof as well, since they're (potentially) stale.